### PR TITLE
TS migration

### PR DESCRIPTION
**Pull request by 0xorial**
I migrated a big part of code to TypeScript. Tests seem to be ok. Minor structural changes, mostly reducing object shape mutations (https://mathiasbynens.be/notes/shapes-ics). Added branding on numerical types.

### CODE REVIEW

**Comment by cmdcolin** [comment]
wow great work! could try and run it through some of our profiling to see if speed is faster/slower/same. i would be curious if the branded number checks have any impact

**Comment by rbuels** [comment]
oh nice, thanks for this!

File: test/lib/fasta/indexedFasta.js [hunk 1]
```diff
@@ -104,7 +104,7 @@ export default class IndexedFasta {
     }
 
     let residues = Buffer.allocUnsafe(readlen)
-    await this.data.read(residues, 0, readlen, position)
+    await this.data.read(headbytes, 0, cramFileDefinitionParser.maxLength, 0)
```
Comment (rbuels) [inline test/lib/fasta/indexedFasta.js:107]: what is this? doesn't seem correct

File: src/io/index.ts [hunk 2]
```diff
@@ -1,16 +1,21 @@
 import url from 'url'
-import RemoteFile from './remoteFile'
-import LocalFile from './localFile'
+import { ensureNotNullish } from '../typescript'
+import { Filehandle } from '../cramFile/filehandle'
+import { LocalFile, RemoteFile } from 'generic-filehandle'
```
Comment (rbuels) [inline src/io/index.ts:4]: nice, been meaning to bring this module up to date to use generic-filehandle

File: src/errors.ts [hunk 3]
```diff
@@ -9,7 +9,6 @@ export class CramMalformedError extends CramError {}
 /**
  * An error caused by attempting to read beyond the end of the defined data.
  */
-export class CramBufferOverrunError extends CramMalformedError {}
```
Comment (rbuels) [inline src/errors.ts:12]: delete its comment too if we really don't need it

**Review by rbuels — COMMENTED** [review]
Just a few nits here and there, code-wise.

I'm a little concerned about the performance implications of all the `assert*` calls, but I haven't tested or benchmarked it. Maybe they're not a big deal

**Comment by cmdcolin** [comment]
timings are pretty similar before and after this PR https://github.com/cmdcolin/jb2profile but a little bit slower.


if we want to go aggressive for performance, it could be worth addressing, but the typescript is a big win. the numbers are about ~10% slower on this branch to do the full jbrowse 2 rendering pipeline (the cram-js portion is a subset of that, so it could be that the cram-js parsing is somewhat larger than 10% slower)


log of curr_cramjs (current release) and next_cramts (this typescript branch) on 20x, 400x, and 800x short reads shows this branch about 10% slower

```


0.02 cram 20x 19
./profile.sh chr22_mask:124000-143000 20x.shortread.cram results/20x-19kb-shortread-cram hg19mod
Benchmark 1: curr_cramjs,20x.shortread.cram
  Time (mean ± σ):      4.608 s ±  0.067 s    [User: 0.388 s, System: 0.124 s]
  Range (min … max):    4.559 s …  4.684 s    3 runs

Benchmark 2: next_cramts,20x.shortread.cram 
  Time (mean ± σ):      4.882 s ±  0.045 s    [User: 0.386 s, System: 0.111 s]
  Range (min … max):    4.831 s …  4.919 s    3 runs

Summary
  'curr_cramjs,20x.shortread.cram' ran
    1.06 ± 0.02 times faster than 'next_cramts,20x.shortread.cram'








0.40 cram 400x 19
./profile.sh chr22_mask:124000-143000 400x.shortread.cram results/400x-19kb-shortread-cram hg19mod
Benchmark 1: curr_cramjs,400x.shortread.cram
  Time (mean ± σ):     22.412 s ±  0.245 s    [User: 0.569 s, System: 0.182 s]
  Range (min … max):   22.192 s … 22.676 s    3 runs

Benchmark 2: next_cramts,400x.shortread.cram
  Time (mean ± σ):     24.631 s ±  0.410 s    [User: 0.562 s, System: 0.173 s]
  Range (min … max):   24.219 s … 25.040 s    3 runs

Summary
  'curr_cramjs,400x.shortread.cram'  ran
    1.10 ± 0.02 times faster than 'next_cramts,400x.shortread.cram"'








0.80 cram 800x 19
./profile.sh chr22_mask:124000-143000 800x.shortread.cram results/800x-19kb-shortread-cram hg19mod
Benchmark 1: curr_cramjs,800x.shortread.cram
  Time (mean ± σ):     41.025 s ±  0.353 s    [User: 0.711 s, System: 0.221 s]
  Range (min … max):   40.651 s … 41.353 s    3 runs

Benchmark 2: next_cramts,800x.shortread.cram
  Time (mean ± σ):     45.014 s ±  1.001 s    [User: 0.703 s, System: 0.239 s]
  Range (min … max):   43.891 s … 45.815 s    3 runs

Summary
  'curr_cramjs,800x.shortread.cram"  ran
    1.10 ± 0.03 times faster than 'next_cramts,800x.shortread.cram'






```

**Comment by cmdcolin** [comment]
overall it is a pretty tiny difference though (think: 2 seconds on a 22 second rendering time) so not very noticeable

File: test/lib/fasta/indexedFasta.js (hunk 1 above)
Comment (0xorial) [reply test/lib/fasta/indexedFasta.js]: wow, I think at some point webstorm went too intelligent on me and screwed up a bunch of .read() function calls around the project. what is a bit strange - this code is in the test folder, but it does not seem to have any impact on tests outcome...

**Comment by 0xorial** [comment]
Regarding the type assertions, I am fairly confident that they will be inlined: https://blog.logrocket.com/how-javascript-works-optimizing-the-v8-compiler-for-efficiency/

Typed additions probably won't though... Also, I was wondering if there is any way to do overflow checks nicer (https://stackoverflow.com/questions/199333/how-do-i-detect-unsigned-integer-overflow)

File: package.json [hunk 4]
```diff
@@ -51,11 +52,16 @@
     "long": "^4.0.0",
     "md5": "^2.2.1",
     "pako": "^1.0.4",
-    "quick-lru": "^2.0.0"
+    "quick-lru": "^6.1.1"
```
Comment (cmdcolin) [inline package.json:55]: might suggest dropping this back down and using @types/quick-lru. the newer version uses too many new js constructs and requires e.g. the transformignorepatterns above which basically implies consumers of cram-js would feel this too

File: .idea/codeStyles/codeStyleConfig.xml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
+  <state>
+    <option name="USE_PER_PROJECT_SETTINGS" value="true" />
```
Comment (cmdcolin) [inline .idea/codeStyles/codeStyleConfig.xml:3]: unless it is quite helpful, maybe can remove .idea folder from pr and add to .gitignore

**Comment by cmdcolin** [comment]
I think we could probably merge this and iterate on it as needed i think. the typescript is great to have

there might be some small changes we could address either in this PR , i added a couple small comments

as far as bounds checking and such, my personal outlook is that it's not necessarily our responsibility to do bounds checking, i expect tools like samtools to do it more than us. it could be superstition but i just think it adds speed slowdowns and js already doesn't really use specific numeric types... if we were to make a wasm version of this, it is probably more natural to actually check for int32 and stuff like that but not as necessary in js

File: package.json (hunk 4 above)
Comment (0xorial) [reply package.json]: v2 are suspiciously not used. updated to v4, which should still be backward compatible, but already have integrated correct typing (v2 was causing typing errors somewhere)

**Comment by cmdcolin** [comment]
@0xorial  I made a modification to this PR that removes the branding aspect here #119 

I think with that change it should be good to go :)

**Comment by cmdcolin** [comment]
CC @rbuels

**Comment by rbuels** [comment]
Sounds good to me

**Comment by cmdcolin** [comment]
super :) i may close this PR and merge mine with credit to @0xorial !

**Comment by cmdcolin** [comment]
or we'll just merge this and then merge mine...looks nicer in the git history

**Pull request merged**